libdnf-plugins/actions/actions.cpp

```diff
@@ -0,0 +1,725 @@
+    std::string pkg_filter;
+    enum class Direction { IN, OUT, ALL } direction;
+    std::string command;
+    std::vector<std::string> args;
+};
+
+
+// Represents one command to run
+struct CommandToRun {
+    [[nodiscard]] bool operator<(const CommandToRun & other) const noexcept;
+    std::string command;
+    std::vector<std::string> args;
+};
+
+
+class Actions : public IPlugin {
+public:
+    Actions(libdnf::Base & base, libdnf::ConfigParser &) : base(base) {}
+    virtual ~Actions() = default;
+
+    APIVersion get_api_version() const noexcept override { return PLUGIN_API_VERSION; }
+
+    const char * get_name() const noexcept override { return PLUGIN_NAME; }
+
+    PluginVersion get_version() const noexcept override { return PLUGIN_VERSION; }
+
+    const char * const * get_attributes() const noexcept override { return attrs; }
+
+    const char * get_attribute(const char * attribute) const noexcept override {
+        for (size_t i = 0; attrs[i]; ++i) {
+            if (std::strcmp(attribute, attrs[i]) == 0) {
+                return attrs_value[i];
+            }
+        }
+        return nullptr;
+    }
+
+    void init() override { parse_action_files(); }
+
+    void pre_base_setup() override { setup(pre_base_setup_actions); }
+
+    void post_base_setup() override { setup(post_base_setup_actions); }
+
+    void pre_transaction(const libdnf::base::Transaction & transaction) override {
+        transaction_impl(transaction, pre_trans_actions);
+    }
+
+    void post_transaction(const libdnf::base::Transaction & transaction) override {
+        transaction_impl(transaction, post_trans_actions);
+    }
+
+private:
+    void parse_action_files();
+    void setup(const std::vector<Action> & trans_actions);
+    void transaction_impl(const libdnf::base::Transaction & transaction, const std::vector<Action> & trans_actions);
+    void execute_command(CommandToRun & command);
+
+    [[nodiscard]] std::pair<std::string, bool> substitute(
+        const libdnf::base::TransactionPackage * trans_pkg,
+        const libdnf::rpm::Package * pkg,
+        std::string_view in,
+        std::filesystem::path file,
+        int line_number);
+
+    [[nodiscard]] std::pair<std::vector<std::string>, bool> substitute_args(
+        const libdnf::base::TransactionPackage * trans_pkg, const libdnf::rpm::Package * pkg, const Action & action);
+
+    void process_command_output_line(std::string_view line);
+
+    libdnf::Base & base;
+
+    // Parsed actions for individual hooks
+    std::vector<Action> pre_base_setup_actions;
+    std::vector<Action> post_base_setup_actions;
+    std::vector<Action> pre_trans_actions;
+    std::vector<Action> post_trans_actions;
+
+    // cache for sharing between pre_transaction and post_transaction hooks
+    bool transaction_cached = false;
+    std::vector<libdnf::base::TransactionPackage> trans_packages;
+    std::map<libdnf::rpm::PackageId, const libdnf::base::TransactionPackage *> pkg_id2trans_pkg;
```

Comment:
Nitpick, but I'd prefer to name this `pkg_id_to_trans_pkg`.

Reply:
OK